v1.1.1/js/Template.js

```diff
@@ -0,0 +1,172 @@
+	modelCopy.getNode = function(id) {
+		return self.nodes[id-1];	
+	}
+
+	// Load template from prototype
+	for (var i = 0; i < prototype.nodes.length; i++) self.nodes.push(new Node(model, Object.assign({}, prototype.nodes[i])));
+	for (var i = 0; i < prototype.edges.length; i++) self.edges.push(new Edge(modelCopy, Object.assign({}, prototype.edges[i])));
+	for (var i = 0; i < prototype.labels.length; i++) self.labels.push(new Label(model, Object.assign({}, prototype.labels[i])));
+
+	// MOUSE.
+	var _listenerMouseUp = subscribe("mouseup", function() {
+		if(self.loopy.mode!=Loopy.MODE_EDIT) return;
+		if(self.loopy.tool!=Loopy.TOOL_TEMPLATE) return;
+
+		// Apply to actual model
+		self.applyTemplate();
+	});
+
+	self.applyTemplate = function() {	
+		// Apply all template components to graph from prototype
+
+		// Keep track of nodes generated in model
+		var generatedNodes = [];
+
+		// Nodes
+		for (var i = 0; i < prototype.nodes.length; i++) {
+			// Create copy of the object
+			var node = Object.assign({}, prototype.nodes[i]);
+
+			// Offset based on selected position
+			node.x = Mouse.x + (node.x || 0);
+			node.y = Mouse.y + (node.y || 0);
+			generatedNodes.push(self.model.addNode(node));
+		}
+
+
+		// Edges
+		for (var i = 0; i < prototype.edges.length; i++) {
+			// Create copy of the object
+			var edge = Object.assign({}, prototype.edges[i]);
+
+			// Re-link edges to corresponding nodes in real graph
+			edge.from = generatedNodes[edge.from - 1].id
+			edge.to = generatedNodes[edge.to - 1].id
+
+			self.model.addEdge(edge);
+		}
+
+		// Labels
+		for (var i = 0; i < prototype.labels.length; i++) {
+			// Create copy of the object
+			var label = Object.assign({}, prototype.labels[i]);
+
+			// Offset based on selected position
+			label.x = Mouse.x + (label.x || 0);
+			label.y = Mouse.y + (label.y || 0);
+			self.model.addLabel(label);
+		}
+
+	};
+
+
+	//////////////////////////////////////
+	// UPDATE & DRAW /////////////////////
+	//////////////////////////////////////
+	
+	// Update
+	self.update = function(speed) {
+		if(self.loopy.mode!=Loopy.MODE_EDIT) return;
+		if(self.loopy.tool!=Loopy.TOOL_TEMPLATE) return;
+
+		// Update all edges then nodes
+		for (var i = 0; i < self.edges.length; i++) self.edges[i].update(speed);
+		for (var i = 0; i < self.nodes.length; i++) self.nodes[i].update(speed);
+	}
+
+	// Draw
+	var _circleRadius = 0;
+	self.draw = function(ctx) {
+		if(self.loopy.mode!=Loopy.MODE_EDIT) return;
+		if(self.loopy.tool!=Loopy.TOOL_TEMPLATE) return;
+		
+		// Save context
+		ctx.save();
+
+		// Apply translation to mouse position
+		ctx.translate(Mouse.x * 2, Mouse.y * 2);
+
+		// Draw labels THEN edges THEN nodes from template
+		for (var i = 0; i < self.labels.length; i++) self.labels[i].draw(ctx);
+		for (var i = 0; i < self.edges.length; i++) self.edges[i].draw(ctx);
+		for (var i = 0; i < self.nodes.length; i++) self.nodes[i].draw(ctx);
+
+		// Restore
+		ctx.restore();
+

```

Comment:
After drawing the template on the canvas, would it be possible to have the selected tool default back to the Pencil tool? I was initially a little worried because I couldn't figure out how to get out of Template mode. I don't really foresee many people dropping down template after template in rapid succession, so I think clearing the selected template and reverting the cursor back to a default tool (pencil feels most natural to me) after placing one would be the best approach.

Additionally, I would like it to dismiss the template if the user clicks off of the drawable canvas (i.e. the sidebar, toolbar, or graph). Right now when you do any of those, the template ghost just sorta chills at the last valid cursor position until you move the mouse back, and it just feels odd to me.